File: templates/python/check.py
import json

class Check:
    @staticmethod
    def initialize(parts):
        Check.parts = parts
        Check.current = None
        Check.part_counter = None

    @staticmethod
    def part():
        if Check.part_counter is None:
            Check.part_counter = 0
        else:
            Check.part_counter += 1
        Check.current = Check.parts[Check.part_counter]
        return Check.current.get('solution', '') != ''

    @staticmethod
    def error(msg):
        if 'errors' not in Check.current:
            Check.current['errors'] = []
        Check.current['errors'].append(msg)

    @staticmethod
    def challenge(x):
        if 'challenge' not in Check.current:
            Check.current['challenge'] = ''
        Check.current['challenge'] += str(x)

    @staticmethod
    def run(example, state, message=None, env={}, clean=lambda x: x):
        s = {}
        s.update(env)
        exec (example, globals(), s)
        for (x,v) in state.items():
            if x not in s:
                Check.error('Ukazi\n{0}\nne nestavijo spremenljivke {1}, a bi jo morali.'.format(example, x))
            elif clean(s[x]) != clean(v):
                Check.error('Ukazi\n{0}\nbi morali nastaviti spremenljivko {1} na {2},\na jo nastavijo na {3}'.format(example, x, v, s[x]))

    @staticmethod
    def equal(example, expected, message=None, clean=lambda x: x, env={}):
        if not message:
            message = 'Ukaz {0} vrne {1!r} namesto {2!r}.'
        local = locals()
        local.update(env)
        answer = eval(example, globals(), local)
        if clean(answer) != clean(expected):
            Check.error(message.format(example, answer, expected))

    @staticmethod
    def num_equal(example, expected, digits=6):
        Check.equal(example, expected, clean=lambda x: round(x, digits))

    @staticmethod
    def summarize():
        for i, part in enumerate(Check.parts):
            if not part['solution']:
                print('Naloga {0} je brez rešitve.'.format(i + 1))
            elif 'errors' in part:
                print('Naloga {0} je napačno rešena.'.format(i + 1))
                print('- {0}'.format("\n- ".join(part['errors'])))
            else:
                print('Naloga {0} je pravilno rešena.'.format(i + 1))

    @staticmethod
    def dump():
        return json.dumps(Check.parts)

File: tomo/templates/r/check.r
check <- list()

check$initialize <- function(parts) {
  check$parts <<- parts
  check$parts$errors <<- list(list())
  check$parts$challenge <<- list(list())
  check$current <<- NA
  check$part.counter <<- NA
}

check$part <- function() {
  if(is.na(check$part.counter)) {
    check$part.counter <<- 1
  } else {
    check$part.counter <<- check$part.counter + 1
  }
  return(check$parts$solution[[check$part.counter]] != "")
}

check$error <- function(msg) {
  check$parts$errors[[check$part.counter]] <<-
    c(check$parts$errors[[check$part.counter]], msg)
}

check$challenge <- function(x, k = NA) {
  pair <- c(toString(k), toString(check$canonize(x)))
  check$parts$challenge[[check$part.counter]] <<-
    c(check$parts$challenge[[check$part.counter]], list(pair))
}

check$run <- function(example, state) {
  # yet to be implemented
}

check$canonize <- function(x, digits = 6) {
  if(typeof(x) == "double" || typeof(x) == "complex") {
    return(round(x, digits))
  } else if(typeof(x) == "complex") {
    return(round(x, digits))
  } else if(typeof(x) == "list") {
    return(lapply(x, function(y) canonize(y, digits)))
  } else {
    return(x)
  }
}

check$compare <- function(example, expected,
                          message = "Ukaz %s vrne %s namesto %s",
                          clean = NA, digits = 6, precision = 1.0e-6,
                          strict_float = FALSE, strict_list = TRUE) {
  example <- substitute(example)
  answer <- try(eval(example), silent = TRUE)
  if(is.na(clean))
    clean <- function(x) check$canonize(x, digits)
  # give a reason for difference like in check.py
  if(!identical(clean(answer), clean(expected))) {
    pretty.print <- function(x) {
      output <- capture.output(
        if(length(dim(x)) <= 1) cat(x) else print(x)
      )
      if(length(output) == 0) {
        return("NULL")
      } else if(length(output) == 1) {
        return(output)
      } else {
        return(paste("    ", c("", output, ""), collapse="\n"))
      }
    }
    check$error(sprintf(message, deparse(example),
                pretty.print(answer), pretty.print(expected)))
  }
}

check$summarize <- function() {
  for(i in 1:nrow(check$parts)) {
    if(check$parts$solution[[i]] == "") {
      cat("Podnaloga", i, "je brez rešitve.\n")
    } else if (length(check$parts$errors[[i]]) > 0) {
      cat("Podnaloga", i, "ni prestala vseh testov.\n")
      cat(paste("- ", check$parts$errors[[i]], "\n", sep = ""), sep="")
    } else {
      cat("Podnaloga", i, "je prestala vse teste.\n")
    }
  }
}

check$dump <- function() {
  return(toJSON(check$parts))
}

